tests/UnifiedSpecTests/Constraint/Matches.php

```diff
@@ -263,6 +268,35 @@ private function assertMatchesOperator(BSONDocument $operator, $actual, string $
             return;
         }
 
+        if ($name === '$$matchAsDocument') {
+            assertInstanceOf(BSONDocument::class, $operator['$$matchAsDocument'], '$$matchAsDocument requires a BSON document');
+            assertIsString($actual, '$$matchAsDocument requires actual value to be a JSON string');
+            assertJson($actual, '$$matchAsDocument requires actual value to be a JSON string');
+
+            /* Note: assertJson() accepts array and scalar values, but the spec
+             * assumes that the JSON string will yield a document. */
+            assertStringStartsWith('{', trim($actual), '$$matchAsDocument requires actual value to be a JSON string denoting an object');
+
+            $actualDocument = Document::fromJSON($actual)->toPHP();
+            $constraint = new Matches($operator['$$matchAsDocument'], $this->entityMap, allowExtraRootKeys: false);
+
+            if (! $constraint->evaluate($actualDocument, '', true)) {
+                self::failAt(sprintf('%s did not match: %s', (new Exporter())->shortenedExport($actual), $constraint->additionalFailureDescription(null)), $keyPath);
+            }
+
+            return;
+        }
+
+        if ($name === '$$matchAsRoot') {
+            $constraint = new Matches($operator['$$matchAsRoot'], $this->entityMap, allowExtraRootKeys: true);
```

Comment:
Why `allowExtraRootKeys` is `true` for `$$matchAsRoot` and `false` for `$$matchAsDocument`?

Reply:
Per the UTF spec, [`$$matchAsDocument`](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.md#matchasdocument) does _not_ allow extra keys in the decoded JSON document.

[`$$matchAsRoot`](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.md#matchasroot) only exists to enable "allow extra keys" behavior when matching.